library/base.go

```diff
@@ -0,0 +1,227 @@
+		log.Fatalln("cannot clean up cgroup dir", fn, err)
+	}
+
+	// clean up temp dir
+	if rp, ok := r.getRunningProcess(puuid); ok {
+		log.Println("cleaning up", rp.tempdir)
+		if err := os.RemoveAll(rp.tempdir); err != nil {
+			log.Fatalln("cannot clean up temp dir", rp.tempdir, err)
+		}
+	} else {
+		log.Println("could not find process", puuid)
+	}
+}
+
+func cleanup(r *RunnerImpl, puuid Process) func() {
+	return func() {
+		r.internalCleanup(puuid)
+	}
+}
+
+func (r *RunnerImpl) addRunningProcess(puuid Process, rp *runningProcess) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.running[puuid] = rp
+}
+
+func (r *RunnerImpl) getRunningProcess(puuid Process) (*runningProcess, bool) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	rp, ok := r.running[puuid]
+	return rp, ok
+}
+
+func (r *RunnerImpl) removeRunningProcess(puuid Process) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	delete(r.running, puuid)
+}
+
+func (r *RunnerImpl) Run(ctx context.Context, client string, cmd string, args ...string) (Process, error) {
+	cgn, ok := r.authz.ClientToCGroup[client]
+	if !ok {
+		return "", errors.New("authorization failed - no cgroup defined for this client")
+	}
+	cg, err := r.authz.getControlGroup(cgn)
+	if err != nil {
+		return "", err
+	}
+
+	puuid := Process(uuid.New().String())
+
+	dir, err := ioutil.TempDir("", string(puuid)+"_*")
+	if err != nil {
+		return "", err
+	}
+
+	os.Chdir(dir)
+
+	if err := r.cgi.MakeCgroup(cg, puuid); err != nil {
+		log.Fatalln("unable to make cgroup:", err)
+	}
+
+	cgDir := fmt.Sprintf("/sys/fs/cgroup/%s", puuid)
+
+	var buf SafeBuffer
+	ps, st, done, err := r.cgi.Run(ctx, r.wrapperPath, cg.SysProcAttr, &buf, cleanup(r, puuid), cgDir, cmd, args...)
+	if err != nil {
+		r.internalCleanup(puuid)
+		return "", err
+	}
+
+	if ps.Process != nil {
+		r.addRunningProcess(puuid, &runningProcess{cmd: ps, output: &buf, tracker: st, pid: ps.Process.Pid, tempdir: dir, done: done})
+		return puuid, nil
+	}
+	r.internalCleanup(puuid)
+	return "", errors.New("process did not launch")
+}
+
+func (r *RunnerImpl) safeGetRunningProcess(p Process) (*runningProcess, error) {
+	rp, ok := r.running[p]
+	if !ok {
+		return nil, errors.New("process not found")
+	}
+	return rp, nil
+}
+
+func (r *RunnerImpl) GetStatus(ctx context.Context, p Process) (Status, int, error) {
+	rp, err := r.safeGetRunningProcess(p)
+	if err != nil {
+		return Unavailable, -1, err
+	}
+	st, ec := rp.tracker.GetStatus()
+	return st, ec, nil
+}
+
+func (r *RunnerImpl) StreamOutput(ctx context.Context, p Process, sender Sender) error {
+	rp, err := r.safeGetRunningProcess(p)
+	if err != nil {
+		return err
+	}
+
+	if st, _ := rp.tracker.GetStatus(); st != Running {
+		return errors.New("process is not running")
+	}
+
+	rp.sender = sender
+
+	log.Println("starting stdout stream")
+
+	// NOTE: we can't do this in a goroutine.
+	// If we return from this method the stream will be closed.
+	done := false
+	reader := rp.output.NewReader()
+	for {
+		select {
```

Comment:
I would expect you to also be checking on the `ctx` that was passed in to this method. It's currently unused.